### PR TITLE
Add specification for `__setitem__`

### PR DESCRIPTION
**Pull request by kgryte**
This PR

-   adds a specification for `__setitem__`.

### CODE REVIEW

File: spec/API_specification/array_object.md [hunk 1]
```diff
@@ -1032,7 +1032,30 @@ Element-wise results must equal the results returned by the equivalent element-w
...
+
+    -   value(s) to set.
+
+```{note}
+
+Setting array values must not affect the data type of `self`.
+
+When `value` is a Python scalar (i.e., `int`, `float`, `bool`), behavior must follow specification guidance on mixing arrays with Python scalars (see {ref}`type-promotion`).
+
+When `value` is an `array` of a different data type than `self`, how values are cast to the data type of `self` is implementation defined.
```
Comment (asmeurer) [inline spec/API_specification/array_object.md:1057]: Why is this? I would have expected the value to just cast to the dtype of self (or error if that isn't allowed).

**Comment by asmeurer** [comment]
Should this mention broadcasting?

File: spec/API_specification/array_object.md (hunk 1 above)
Comment (kgryte) [reply spec/API_specification/array_object.md:1057]: I had in mind how `int <-> uint` conversions would work for overflow/underflow. Do values wrap, error when exceeding bounds, or something else? When converting from a float to an int, are values truncated, trigger an exception, etc, given that how arrays can handle mixed kind ops is implementation defined.

**Comment by kgryte** [comment]
@asmeurer Yes, I will add a note that `value` must be broadcast compatible with `self`. Thanks!

**Comment by asmeurer** [comment]
It should be broadcast compatible with `self[key]`.

File: spec/API_specification/array_object.md (hunk 1 above)
Comment (rgommers) [reply spec/API_specification/array_object.md:1057]: I agree that this is hard to make uniform. Example:
```
In [5]: x = np.linspace(0, 1)

In [6]: x[:2] = np.array([1, 1+1.j])
ComplexWarning: Casting complex values to real discards the imaginary part

In [7]: x[:4] = 1+1.j
ComplexWarning: Casting complex values to real discards the imaginary part
```
I'd hope that that errors in other implementations.

Comment (kgryte) [reply spec/API_specification/array_object.md:1057]: Looks like NumPy is following C [semantics](https://en.cppreference.com/w/c/language/conversion) in the above example. But certainly, other implementations should be free to disallow and raise an exception.

**Review by rgommers — APPROVED** [review]
LGTM. This looks relatively straightforward and was a clear gap, so merging. Thanks @kgryte and @asmeurer

**Pull request merged**